### PR TITLE
feat(ts): Add support for multiline nodes in get_node_text

### PR DESCRIPTION
**Pull request by mfussenegger**
Based on https://github.com/neovim/neovim/pull/14445

This extends `vim.treesitter.query.get_node_text` to return the text
that spans a node's range even if start_row ~= end_row.

### CODE REVIEW

File: runtime/lua/vim/treesitter/query.lua [hunk 1]
```diff
@@ -175,11 +173,22 @@ function M.get_node_text(node, source)
...
-    return string.sub(line, start_col+1, end_col)
+    lines[1] = string.sub(lines[1], start_col + 1)
+
+    local end_index = end_col
+    if #lines == 1 then
+      end_index = end_col - start_col
+    end
+    lines[#lines] = string.sub(lines[#lines], 1, end_index)
+
+    return table.concat(lines, "\n")
```
Comment (mfussenegger) [inline runtime/lua/vim/treesitter/query.lua:191]: Would it make sense to have a `nvim_buf_get_text` and use that here?

File: runtime/lua/vim/treesitter/query.lua [hunk 2]
```diff
@@ -175,11 +173,22 @@ function M.get_node_text(node, source)
   local end_row, end_col, end_byte = node:end_()
 
   if type(source) == "number" then
-    if start_row ~= end_row then
-      return nil
+    local lines
+    if end_col == 0 then
+      lines = a.nvim_buf_get_lines(source, start_row, end_row, true)
```
Comment (stsewd) [inline runtime/lua/vim/treesitter/query.lua:178]: You'll need to check for start/end row not being larger than the number of lines of the buffer, similar to https://github.com/neovim/neovim/pull/15030/

Comment (mfussenegger) [reply runtime/lua/vim/treesitter/query.lua:178]: Do you have an example where this happens? I'd like to create a test case for it

Comment (stsewd) [reply runtime/lua/vim/treesitter/query.lua:178]: he, this is kind of hard to replicate with the c parser, as it needs to have a node that includes the EOF mark at the start and/or end of de node. I was using this to replicate cases like this


```lua
local parsers = require "nvim-treesitter.parsers"
local query = require "vim.treesitter.query"

local function print(str)
  vim.api.nvim_err_writeln(vim.inspect(str))
end

local function show(node)
  print { node:start() }
  print { node:end_() }
  print(query.get_node_text(node, 0))
end

local lang = "ruby"

local parser = parsers.get_parser(0, lang)
local tree = parser:parse()[1]
local node = tree:root()
local q = query.parse_query(lang, "(heredoc_end) @this")
for pattern, match, metadata in q:iter_matches(node, 0, 1, 4) do
  for id, n in pairs(match) do
    show(n)
  end
end
```

Executed while having this file open

```ruby
def run
  a = <<~E
end
```

I think is easier to test this mocking the node, but don't know if that's possible to do in those tests.

Comment (mfussenegger) [reply runtime/lua/vim/treesitter/query.lua:182]: Thanks. I've added a test case

File: runtime/lua/vim/treesitter/query.lua (hunk 1 above)
Comment (mfussenegger) [reply runtime/lua/vim/treesitter/query.lua]: Turns out there is already a PR for it - https://github.com/neovim/neovim/pull/15181   once it is merged I can update the PR here to use it.

**Review by vigoux — COMMENTED** [review]
A little nitpick I have: it seems (from memory) that the multiline `match?` would only work if the operator is actually `lua-match?`, which might be a problem.

@bfredl do you remember anything about this ?

**Review by vigoux — APPROVED** [review]

**Pull request merged**